tsfc/loopy.py

```diff
@@ -0,0 +1,340 @@
+def statement_returnaccumulate(leaf, ctx):
+    lhs = expression(leaf.variable, ctx)
+    rhs = lhs + expression(leaf.indexsum.children[0], ctx)
+    return [lp.Assignment(lhs, rhs, within_inames=ctx.active_inames())]
+
+
+@statement.register(imp.Evaluate)
+def statement_evaluate(leaf, ctx):
+    expr = leaf.expression
+    if isinstance(expr, gem.ListTensor):
+        ops = []
+        var = ctx.pymbolic_variable(expr)
+        index = ()
+        if isinstance(var, p.Subscript):
+            var, index = var.aggregate, var.index_tuple
+        for multiindex, value in numpy.ndenumerate(expr.array):
+            ops.append(lp.Assignment(p.Subscript(var, index + multiindex), expression(value, ctx), within_inames=ctx.active_inames()))
+        return ops
+    elif isinstance(expr, gem.Constant):
+        return []
+    else:
+        return [lp.Assignment(ctx.pymbolic_variable(expr), expression(expr, ctx, top=True), within_inames=ctx.active_inames())]
+
+
+def expression(expr, ctx, top=False):
+    """Translates GEM expression into a pymbolic expression
+
+    :arg expr: GEM expression
+    :arg ctx: miscellaneous code generation data
+    :arg top: do not generate temporary reference for the root node
+    :returns: pymbolic expression
+    """
+    if not top and expr in ctx.gem_to_pymbolic:
+        return ctx.pymbolic_variable(expr)
+    else:
+        return _expression(expr, ctx)
+
+
+@singledispatch
+def _expression(expr, parameters):
+    raise AssertionError("cannot generate expression from %s" % type(expr))
+
+
+@_expression.register(gem.Failure)
+def _expression_failure(expr, parameters):
+    raise expr.exception
+
+
+@_expression.register(gem.Product)
+def _expression_product(expr, ctx):
+    return p.Product(tuple(expression(c, ctx) for c in expr.children))
+
+
+@_expression.register(gem.Sum)
+def _expression_sum(expr, ctx):
+    return p.Sum(tuple(expression(c, ctx) for c in expr.children))
+
+
+@_expression.register(gem.Division)
+def _expression_division(expr, ctx):
+    return p.Quotient(*(expression(c, ctx) for c in expr.children))
+
+
+@_expression.register(gem.Power)
+def _expression_power(expr, ctx):
+    return p.Power(*(expression(c, ctx) for c in expr.children))
+
+
+@_expression.register(gem.MathFunction)
+def _expression_mathfunction(expr, ctx):
+    name_map = {
+        'abs': 'abs',
+        'ln': 'log'
+    }
+    name = name_map.get(expr.name, expr.name)
+    if name == 'jn':
+        nu, arg = expr.children
+        if nu == gem.Zero():
+            return p.Variable("j0")(expression(arg, ctx))
+        elif nu == gem.one:
+            return p.Variable("j1")(expression(arg, ctx))
+    if name == 'yn':
+        nu, arg = expr.children
+        if nu == gem.Zero():
+            return p.Variable("y0")(expression(arg, ctx))
+        elif nu == gem.one:
+            return p.Variable("y1")(expression(arg, ctx))
+    return p.Variable(name)(*[expression(c, ctx) for c in expr.children])
```

Comment:
This function needs updating... see current `tsfc/coffee.py`.

Reply:
Done -  but need testing on the complex branch.